File: 2025-12/src/day_08/boj_2312/Main.java
package day_08.boj_2312;

import java.io.*;
import java.util.*;

public class Main {

    static final int MAX = 100_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        List<Integer> primes = getPrimes(sieve());

        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int n = Integer.parseInt(br.readLine());

            for (int p : primes) {
                if (p * p > n) break;  // prime의 제곱보다 작으면 탐색 종료

                int cnt = 0;
                while (n % p == 0) {
                    n /= p;
                    cnt++;
                }

                if (cnt > 0) sb.append(p).append(" ").append(cnt).append("\n");
            }

            if (n > 1) sb.append(n).append(" 1\n");  // N이 처음부터 소수인 경우와 합성수를 소인수분해한 후 소수가 남은 경우
        }

        System.out.println(sb);
    }

    static boolean[] sieve() {
        boolean[] isPrime = new boolean[1 + MAX];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;

        for (int i = 2; i * i <= MAX; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= MAX; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        return isPrime;
    }

    static List<Integer> getPrimes(boolean[] isPrime) {
        List<Integer> primes = new ArrayList<>();

        for (int i = 2; i < isPrime.length; i++) {
            if (isPrime[i]) primes.add(i);
        }

        return primes;
    }
}

File: 2025-12/src/day_08/boj_2312/Main.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int MAX = 100000;
bool isPrime[1 + MAX];
vector<int> primes;

void sieve() {
    fill(isPrime, isPrime + MAX + 1, true);
    isPrime[0] = isPrime[1] = false;

    for (int i = 2; i * i <= MAX; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j <= MAX; j += i) {
                isPrime[j] = false;
            }
        }
    }

    for (int i = 2; i <= MAX; i++) {
        if (isPrime[i]) primes.push_back(i);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    sieve();

    int t;
    cin >> t;

    while (t--) {
        int n;
        cin >> n;

        for (int p : primes) {
            if (p * p > n) break;  // p의 제곱보다 작으면 탐색 종료

            int cnt = 0;
            while (n % p == 0) {
                n /= p;
                cnt++;
            }

            if (cnt > 0) {
                cout << p << ' ' << cnt << '\n';
            }
        }

        if (n > 1) {
            cout << n << ' ' << 1 << '\n';  // n이 처음부터 소수인 경우와 합성수를 소인수분해한 후 소수가 남은 경우
        }
    }
}
